File: bin/schemas/canon_schema.py
"""Continuity layer: facts extracted from the manuscript, and the
contradictions found among them.

## Why this is a layer and not a phase

Developmental editing precedes line editing because structural revision
destroys line work — no point polishing a sentence in a scene you'll cut.
Continuity isn't destroyed that way. Cut a chapter and the canon simply
loses those facts; the ones that remain are still true. So continuity
runs alongside whatever phase you're in rather than waiting its turn.

## Why extraction carries no judgment

Extraction is deliberately dumb: it records what the text asserts, with
provenance, and nothing else. A `Fact` has no `note`, no `severity`, no
`concern` field — there is nowhere for an opinion to live, and
`validate_observations` rejects unknown keys outright. This is
structural, not stylistic: an extractor that can editorialize will start
flagging contradictions it half-noticed, and then contradiction detection
is happening in two places with two standards.

Judgment happens once, in reconciliation, and only on collisions a script
already found mechanically.
"""

from __future__ import annotations

import re

from schemas.findings_schema import SEVERITIES, STATUSES

ENTITY_TYPES = {
    "character",
    "place",
    "object",
    "organization",
    "event",
    "world_rule",
}

# Where the assertion comes from. This matters in fiction: two narration
# facts that disagree is an error, but a character stating something the
# narration contradicts may just be a character who is lying or wrong.
SOURCES = {"narration", "dialogue", "character_thought"}

# "her green eyes" is explicit; "she reached the top shelf easily" implies
# height. Implied facts shouldn't drive hard contradictions on their own.
CONFIDENCES = {"explicit", "implied"}

CONTRADICTION_KINDS = {
    "contradiction",  # two assertions that cannot both be true
    "unverified",  # looks wrong, but needs the author -- lying character,
    # unreliable narrator, or possible in-world explanation
}

CONTINUITY_CATEGORIES = {
    "character_attribute",
    "timeline",
    "geography",
    "world_rule",
    "naming",
    "relationship",
    "object",
}

FACT_REQUIRED_KEYS = {
    "id",
    "entity",
    "entity_type",
    "attribute",
    "value",
    "excerpt",
    "location",
    "source",
    "confidence",
}
# Intentionally empty. Adding an optional key here is how this schema stops
# being judgment-free -- think hard before doing it.
FACT_OPTIONAL_KEYS: set = set()

FACT_ID_PATTERN = re.compile(r"^fact-[a-z0-9_]+-\d{3}$")
CONTRADICTION_ID_PATTERN = re.compile(r"^cont-\d{3}$")


def _check_fact(fact: dict, index: int, errors: list, seen_ids: set) -> None:
    prefix = f"facts[{index}]"
    if not isinstance(fact, dict):
        errors.append(f"{prefix}: not an object")
        return

    keys = set(fact)
    missing = FACT_REQUIRED_KEYS - keys
    if missing:
        errors.append(f"{prefix}: missing keys {sorted(missing)}")

    # The point of the whole schema: no room for an opinion.
    extra = keys - FACT_REQUIRED_KEYS - FACT_OPTIONAL_KEYS
    if extra:
        errors.append(
            f"{prefix}: unexpected keys {sorted(extra)} — extraction records facts, not judgments"
        )

    fact_id = fact.get("id")
    if not isinstance(fact_id, str) or not FACT_ID_PATTERN.match(fact_id or ""):
        errors.append(
            f"{prefix}: id {fact_id!r} does not match {FACT_ID_PATTERN.pattern}"
        )
    elif fact_id in seen_ids:
        errors.append(f"{prefix}: duplicate id {fact_id!r}")
    else:
        seen_ids.add(fact_id)

    if fact.get("entity_type") not in ENTITY_TYPES:
        errors.append(
            f"{prefix}: entity_type {fact.get('entity_type')!r} not in {sorted(ENTITY_TYPES)}"
        )

    if fact.get("source") not in SOURCES:
        errors.append(
            f"{prefix}: source {fact.get('source')!r} not in {sorted(SOURCES)}"
        )

    if fact.get("confidence") not in CONFIDENCES:
        errors.append(
            f"{prefix}: confidence {fact.get('confidence')!r} not in {sorted(CONFIDENCES)}"
        )

    for key in ("entity", "attribute", "value", "excerpt", "location"):
        if key in keys and not str(fact.get(key) or "").strip():
            errors.append(f"{prefix}: missing/empty {key!r}")


def validate_observations(report: dict) -> list:
    """Validate one chapter's extracted facts."""
    errors: list = []
    if not isinstance(report, dict):
        return ["observations file is not a JSON object"]

    if not report.get("chapter"):
        errors.append("missing/empty 'chapter'")

    # Recorded so the skill can skip re-extracting unchanged chapters.
    if not report.get("chapter_sha256"):
        errors.append("missing/empty 'chapter_sha256'")

    facts = report.get("facts")
    if not isinstance(facts, list):
        errors.append("'facts' is not a list")
        return errors

    seen_ids: set = set()
    for i, fact in enumerate(facts):
        _check_fact(fact, i, errors, seen_ids)
    return errors


def validate_continuity_report(report: dict) -> list:
    """Validate the adjudicated contradictions."""
    errors: list = []
    if not isinstance(report, dict):
        return ["continuity report is not a JSON object"]

    contradictions = report.get("contradictions")
    if not isinstance(contradictions, list):
        errors.append("'contradictions' is not a list")
        return errors

    seen_ids: set = set()
    for i, item in enumerate(contradictions):
        prefix = f"contradictions[{i}]"
        if not isinstance(item, dict):
            errors.append(f"{prefix}: not an object")
            continue

        item_id = item.get("id")
        if not isinstance(item_id, str) or not CONTRADICTION_ID_PATTERN.match(
            item_id or ""
        ):
            errors.append(
                f"{prefix}: id {item_id!r} does not match {CONTRADICTION_ID_PATTERN.pattern}"
            )
        elif item_id in seen_ids:
            errors.append(f"{prefix}: duplicate id {item_id!r}")
        else:
            seen_ids.add(item_id)

        if item.get("status") not in STATUSES:
            errors.append(
                f"{prefix}: status {item.get('status')!r} not in {sorted(STATUSES)}"
            )

        if item.get("kind") not in CONTRADICTION_KINDS:
            errors.append(
                f"{prefix}: kind {item.get('kind')!r} not in {sorted(CONTRADICTION_KINDS)}"
            )

        if item.get("category") not in CONTINUITY_CATEGORIES:
            errors.append(
                f"{prefix}: category {item.get('category')!r} not in {sorted(CONTINUITY_CATEGORIES)}"
            )

        if item.get("severity") not in SEVERITIES:
            errors.append(
                f"{prefix}: severity {item.get('severity')!r} not in {sorted(SEVERITIES)}"
            )

        fact_ids = item.get("fact_ids")
        if not isinstance(fact_ids, list) or len(fact_ids) < 2:
            errors.append(
                f"{prefix}: 'fact_ids' must list at least the two conflicting facts"
            )

        if not str(item.get("note") or "").strip():
            errors.append(f"{prefix}: missing/empty 'note'")

    return errors

File: go/internal/schemas/canon_schema.go
package schemas

import (
	"fmt"
	"regexp"
)

// Confidences and ContradictionKinds are universal across domains, same
// split as canon_schema.py: "her green eyes" is explicit; "she reached
// the top shelf easily" implies height. Implied facts shouldn't drive
// hard contradictions on their own.
var Confidences = map[string]bool{"explicit": true, "implied": true}

var ContradictionKinds = map[string]bool{
	"contradiction": true, // two assertions that cannot both be true
	"unverified":    true, // looks wrong, but needs the author -- lying
	// character, unreliable narrator, or possible in-world explanation
}

// factRequiredKeys is deliberately exhaustive with no optional
// extension point -- see canon_schema.py's FACT_OPTIONAL_KEYS comment
// for why: adding an optional key here is how this schema stops being
// judgment-free.
var factRequiredKeys = map[string]bool{
	"id": true, "entity": true, "entity_type": true, "attribute": true,
	"value": true, "excerpt": true, "location": true, "source": true,
	"confidence": true,
}

var factIDPattern = regexp.MustCompile(`^fact-[a-z0-9_]+-\d{3}$`)
var contradictionIDPattern = regexp.MustCompile(`^cont-\d{3}$`)

func checkFact(factRaw interface{}, entityTypes, sources map[string]bool, index int, seenIDs map[string]bool) []string {
	prefix := fmt.Sprintf("facts[%d]", index)
	fact, ok := asObject(factRaw)
	if !ok {
		return []string{prefix + ": not an object"}
	}

	var errors []string

	var missing []string
	for k := range factRequiredKeys {
		if _, present := fact[k]; !present {
			missing = append(missing, k)
		}
	}
	if len(missing) > 0 {
		errors = append(errors, fmt.Sprintf("%s: missing keys %s", prefix, pyList(missing)))
	}

	// The point of the whole schema: no room for an opinion.
	var extra []string
	for k := range fact {
		if !factRequiredKeys[k] {
			extra = append(extra, k)
		}
	}
	if len(extra) > 0 {
		errors = append(errors, fmt.Sprintf("%s: unexpected keys %s — extraction records facts, not judgments", prefix, pyList(extra)))
	}

	id := asString(fact["id"])
	if id == "" || !factIDPattern.MatchString(id) {
		errors = append(errors, fmt.Sprintf("%s: id %s does not match %s", prefix, pyRepr(fact["id"]), factIDPattern.String()))
	} else if seenIDs[id] {
		errors = append(errors, fmt.Sprintf("%s: duplicate id %s", prefix, pyRepr(id)))
	} else {
		seenIDs[id] = true
	}

	entityType := asString(fact["entity_type"])
	if !entityTypes[entityType] {
		errors = append(errors, fmt.Sprintf("%s: entity_type %s not in %s", prefix, pyRepr(fact["entity_type"]), pySetRepr(entityTypes)))
	}

	source := asString(fact["source"])
	if !sources[source] {
		errors = append(errors, fmt.Sprintf("%s: source %s not in %s", prefix, pyRepr(fact["source"]), pySetRepr(sources)))
	}

	confidence := asString(fact["confidence"])
	if !Confidences[confidence] {
		errors = append(errors, fmt.Sprintf("%s: confidence %s not in %s", prefix, pyRepr(fact["confidence"]), pySetRepr(Confidences)))
	}

	for _, key := range []string{"entity", "attribute", "value", "excerpt", "location"} {
		if _, present := fact[key]; present && isBlank(fact[key]) {
			errors = append(errors, fmt.Sprintf("%s: missing/empty %s", prefix, pyRepr(key)))
		}
	}

	return errors
}

// ValidateObservations validates one section's extracted facts.
// `entityTypes`/`sources` come from the manuscript's domain config.
// Mirrors canon_schema.py's validate_observations.
func ValidateObservations(reportRaw interface{}, entityTypes, sources map[string]bool) []string {
	report, ok := asObject(reportRaw)
	if !ok {
		return []string{"observations file is not a JSON object"}
	}

	var errors []string
	if isBlank(report["section"]) {
		errors = append(errors, "missing/empty 'section'")
	}
	// Recorded so the skill can skip re-extracting unchanged sections.
	if isBlank(report["section_sha256"]) {
		errors = append(errors, "missing/empty 'section_sha256'")
	}

	factsRaw, ok := report["facts"].([]interface{})
	if !ok {
		return append(errors, "'facts' is not a list")
	}

	seenIDs := map[string]bool{}
	for i, f := range factsRaw {
		errors = append(errors, checkFact(f, entityTypes, sources, i, seenIDs)...)
	}
	return errors
}

// ValidateContinuityReport validates adjudicated contradictions.
// `categories` is the domain's continuity-category set. Mirrors
// canon_schema.py's validate_continuity_report.
func ValidateContinuityReport(reportRaw interface{}, categories map[string]bool) []string {
	report, ok := asObject(reportRaw)
	if !ok {
		return []string{"continuity report is not a JSON object"}
	}

	contradictionsRaw, ok := report["contradictions"].([]interface{})
	if !ok {
		return []string{"'contradictions' is not a list"}
	}

	var errors []string
	seenIDs := map[string]bool{}
	for i, itemRaw := range contradictionsRaw {
		prefix := fmt.Sprintf("contradictions[%d]", i)
		item, ok := asObject(itemRaw)
		if !ok {
			errors = append(errors, prefix+": not an object")
			continue
		}

		id := asString(item["id"])
		if id == "" || !contradictionIDPattern.MatchString(id) {
			errors = append(errors, fmt.Sprintf("%s: id %s does not match %s", prefix, pyRepr(item["id"]), contradictionIDPattern.String()))
		} else if seenIDs[id] {
			errors = append(errors, fmt.Sprintf("%s: duplicate id %s", prefix, pyRepr(id)))
		} else {
			seenIDs[id] = true
		}

		status := asString(item["status"])
		if !Statuses[status] {
			errors = append(errors, fmt.Sprintf("%s: status %s not in %s", prefix, pyRepr(item["status"]), pySetRepr(Statuses)))
		}

		kind := asString(item["kind"])
		if !ContradictionKinds[kind] {
			errors = append(errors, fmt.Sprintf("%s: kind %s not in %s", prefix, pyRepr(item["kind"]), pySetRepr(ContradictionKinds)))
		}

		category := asString(item["category"])
		if !categories[category] {
			errors = append(errors, fmt.Sprintf("%s: category %s not in %s", prefix, pyRepr(item["category"]), pySetRepr(categories)))
		}

		severity := asString(item["severity"])
		if !Severities[severity] {
			errors = append(errors, fmt.Sprintf("%s: severity %s not in %s", prefix, pyRepr(item["severity"]), pySetRepr(Severities)))
		}

		factIDs, ok := item["fact_ids"].([]interface{})
		if !ok || len(factIDs) < 2 {
			errors = append(errors, fmt.Sprintf("%s: 'fact_ids' must list at least the two conflicting facts", prefix))
		}

		if isBlank(item["note"]) {
			errors = append(errors, fmt.Sprintf("%s: missing/empty 'note'", prefix))
		}
	}
	return errors
}
